### PR TITLE
Implemented Kafka client

### PR DESCRIPTION
**Pull request by jeongkyun-oh**
## Proposed changes

- Implemented Kafka client
- `Publish`, `CreateTopic`, `DeleteTopic`, `ListTopics` are implemented and used later on.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: datasync/chaindatafetcher/kafka/kafka.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	}, nil
+}
+
+func (k *Kafka) CreateTopic(topic string) error {
+	err := k.admin.CreateTopic(topic, &sarama.TopicDetail{
+		NumPartitions:     k.config.partitions,
+		ReplicationFactor: k.config.replicas,
+	}, false)
+
+	return err
```
Comment (ehnuje) [inline datasync/chaindatafetcher/kafka/kafka.go:61]:
```suggestion
	return k.admin.CreateTopic(topic, &sarama.TopicDetail{
		NumPartitions:     k.config.partitions,
		ReplicationFactor: k.config.replicas,
	}, false)
```

File: datasync/chaindatafetcher/kafka/kafka.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	}
+	item.Value = sarama.StringEncoder(data)
+
+	_, _, err = k.producer.SendMessage(item)
+	return err
+}
+
+func newSyncProducer(brokers []string) (sarama.SyncProducer, error) {
+	config := sarama.NewConfig()
+	config.Producer.Return.Successes = true
```
Comment (ehnuje) [inline datasync/chaindatafetcher/kafka/kafka.go:81]: Setting configuration in a function looks weird. If it's a default value, please make another `NewConfig` function for your kafka package and set the default value inside the new `NewConfig` function.

File: datasync/chaindatafetcher/kafka/kafka.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	// TODO-ChainDataFetcher add more configuration if necessary
+	return sarama.NewSyncProducer(brokers, config)
+}
+
+func newClusterAdmin(brokers []string) (sarama.ClusterAdmin, error) {
+	config := sarama.NewConfig()
+	config.Version = sarama.MaxVersion
```
Comment (ehnuje) [inline datasync/chaindatafetcher/kafka/kafka.go:88]: Same comment here.

File: datasync/chaindatafetcher/kafka/kafka_test.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	consumerConfig := sarama.NewConfig()
+	consumer, err := sarama.NewConsumer(config.brokers, consumerConfig)
+	s.NoError(err)
+	s.consumer = consumer
+	s.topic = "test-topic"
+}
+
+func (s *KafkaSuite) TestKafka_CreateAndDeleteTopic() {
+	// no topic to be deleted
+	err := s.kfk.admin.DeleteTopic(s.topic)
```
Comment (ehnuje) [inline datasync/chaindatafetcher/kafka/kafka_test.go:53]: I understand that `DeleteTopic` only receives `topic` string as its parameter and hence is directly called like `kfk.admin.DeleteTopic`, it looses its consistency since `CreateTopic` is called from `Kafka` struct and `DeleteTopic` is called from `Kafka.admin`.

File: datasync/chaindatafetcher/kafka/kafka_test.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	s.Error(err)
+	s.True(strings.Contains(err.Error(), sarama.ErrUnknownTopicOrPartition.Error()))
```
Comment (ehnuje) [inline datasync/chaindatafetcher/kafka/kafka_test.go:55]: Can't we do this like below?
```suggestion
	s.Equal(err, sarama.ErrUnknownTopicOrPartition)
```

File: datasync/chaindatafetcher/kafka/kafka_test.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+	// failed to create a duplicated topic
+	err = s.kfk.CreateTopic(s.topic)
+	s.Error(err)
+	s.True(strings.Contains(err.Error(), sarama.ErrTopicAlreadyExists.Error()))
+
+	// deleted a topic successfully
+	s.Nil(s.kfk.admin.DeleteTopic(s.topic))
+
+	topics, err := s.kfk.admin.ListTopics()
```
Comment (ehnuje) [inline datasync/chaindatafetcher/kafka/kafka_test.go:69]: Same comment for `ListTopics`

File: datasync/chaindatafetcher/kafka/kafka_test.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+	if _, exist := topics[s.topic]; exist {
+		s.Fail("topic must not exist")
+	}
+}
+
+type kafkaData struct {
+	Data []byte `json:"data"`
+}
+
+func createRandomData(size int) []byte {
```
Comment (ehnuje) [inline datasync/chaindatafetcher/kafka/kafka_test.go:79]: Please check `common.MakeRandomBytes` =)

File: datasync/chaindatafetcher/kafka/kafka.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+	}
+
+	data, err := json.Marshal(msg)
+	if err != nil {
+		return err
+	}
+	item.Value = sarama.StringEncoder(data)
+
+	_, _, err = k.producer.SendMessage(item)
+	return err
```
Comment (winnie-byun) [inline datasync/chaindatafetcher/kafka/kafka.go:76]:
```suggestion
	data, err := json.Marshal(msg)
	if err != nil {
		return err
	}

	_, _, err = k.producer.SendMessage(&sarama.ProducerMessage{
		Topic: topic,
		Value: sarama.StringEncoder(data),
	})

	return err
```

File: datasync/chaindatafetcher/kafka/config.go [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+type KafkaConfig struct {
+	brokers    []string // brokers is a list of broker URLs.
+	partitions int32    // partitions is the number of partitions of a topic.
+	replicas   int16    // replicas is a replication factor of kafka settings. This is the number of the replicated partitions in the kafka cluster.
+}
+
+var DefaultKafkaConfig = &KafkaConfig{
+	partitions: int32(1),
+	replicas:   int16(1),
+}
```
Comment (winnie-byun) [inline datasync/chaindatafetcher/kafka/config.go:28]:
```suggestion
func GetDefaultKafkaConfig() *KafkaConfig {
	return &KafkaConfig{
		partitions: 1,
		replicas:   1,
	}
}
```
It would be best to set these values as const. However, because it is unable to set struct as const, how about declaring it as Get function?

File: datasync/chaindatafetcher/kafka/kafka.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+func newSyncProducer(brokers []string) (sarama.SyncProducer, error) {
+	config := sarama.NewConfig()
+	config.Producer.Return.Successes = true
+	// TODO-ChainDataFetcher add more configuration if necessary
+	return sarama.NewSyncProducer(brokers, config)
```
Comment (winnie-byun) [inline datasync/chaindatafetcher/kafka/kafka.go:83]: In the comment of `sarama.SyncProducer` and `sarama.ClusterAdmin `, it says to call `Close()` to avoid leaks. Maybe we should set `Close()` for `Kafka`?

File: datasync/chaindatafetcher/kafka/kafka.go (hunk 1 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/kafka.go]: Applied. 146d2e0230d1882d25d3237c90b324bbc301912d

File: datasync/chaindatafetcher/kafka/kafka_test.go (hunk 7 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/kafka_test.go]: Applied. a029affb0ec88474ea6cd5f5b35b2ea70dca1738

File: datasync/chaindatafetcher/kafka/kafka_test.go (hunk 5 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/kafka_test.go:54]: There is a wrapper function to make `TopicError`. I think the comparison of strings is enough.

File: datasync/chaindatafetcher/kafka/kafka.go (hunk 2 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/kafka.go]: Applied. eb627a05af8dfe3db52342bf37c1b9d9e35f7cd8

File: datasync/chaindatafetcher/kafka/kafka.go (hunk 3 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/kafka.go]: Applied. eb627a05af8dfe3db52342bf37c1b9d9e35f7cd8

File: datasync/chaindatafetcher/kafka/kafka_test.go (hunk 4 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/kafka_test.go]: Alright. I added the methods again for consistency. a788395e18ab2cf5af6581156f53914edce77e84

File: datasync/chaindatafetcher/kafka/kafka_test.go (hunk 6 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/kafka_test.go]: Added. a788395e18ab2cf5af6581156f53914edce77e84

File: datasync/chaindatafetcher/kafka/config.go [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+import "github.com/Shopify/sarama"
+
+type KafkaConfig struct {
+	saramaConfig *sarama.Config // kafka client configurations.
+	brokers      []string       // brokers is a list of broker URLs.
+	partitions   int32          // partitions is the number of partitions of a topic.
+	replicas     int16          // replicas is a replication factor of kafka settings. This is the number of the replicated partitions in the kafka cluster.
+}
+
+func GetDefaultKafkaConfig() *KafkaConfig {
```
Comment (ehnuje) [inline datasync/chaindatafetcher/kafka/config.go:28]: Do we need this as a public function?

Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/config.go]: Nope. Modified. db58f581e1c1544aa3af5572887b8808c3a7f6bf

File: datasync/chaindatafetcher/kafka/kafka.go (hunk 10 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/kafka.go]: Right. Added `Close()` function. 031a754788921c53176a44772de3470b5389e4dc

File: datasync/chaindatafetcher/kafka/config.go (hunk 9 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/config.go]: Updated it here. 89e762171cb733dc1bf6d942cb4d033cfcf5db7f

File: datasync/chaindatafetcher/kafka/kafka.go (hunk 8 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/kafka.go]: Looks better. Thanks. 2dbe802b957c9625754e9b6fe671797c34aba288

**Review by winnie-byun — APPROVED** [review]

**Review by aidan-kwon — APPROVED** [review]

File: build/ci.go [hunk 12]
```diff
@@ -347,6 +347,7 @@ func doFmt(cmdline []string) {
 		"--tests",
 		"--disable-all",
 		"--enable=gofmt",
+		"--timeout=2m",
```
Comment (ehnuje) [inline build/ci.go:350]: Is this related to this PR?

Comment (jeongkyun-oh) [reply build/ci.go:350]: Without this, the linter is failed with timeout error.

**Review by ehnuje — APPROVED** [review]

File: datasync/chaindatafetcher/kafka/config.go [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+// along with the klaytn library. If not, see <http://www.gnu.org/licenses/>.
+
+package kafka
+
+import "github.com/Shopify/sarama"
+
+type KafkaConfig struct {
+	saramaConfig *sarama.Config // kafka client configurations.
+	brokers      []string       // brokers is a list of broker URLs.
+	partitions   int32          // partitions is the number of partitions of a topic.
```
Comment (yoomee1313) [inline datasync/chaindatafetcher/kafka/config.go:24]: To make sure what I understand is correct, partition is a unit of distributed processing and replica is a unit of replication, right?

Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kafka/config.go:24]: So, a topic is kinda message queue which consists of several partitions to process message in parallel. The replica(= replication factor) is how many partitions are copied to other brokers.

For example, Broker1 is a leader for topic1-partition1 and the other brokers have the replicated t1-p1. 

image

Comment (yoomee1313) [reply datasync/chaindatafetcher/kafka/config.go:24]: Fully Understood :)

**Review by yoomee1313 — APPROVED** [review]

**Review by KimKyungup — APPROVED** [review]
